docs/modules/ROOT/pages/adr/0030-function-revisions.adoc

```diff
@@ -0,0 +1,123 @@
+
+Then it can directly be referenced in the composition:
+
+[source, yaml]
+----
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
+metadata:
+  name: vshnkeycloak.vshn.appcat.vshn.io
+  labels:
+    metadata.appcat.vshn.io/revision: v4.120.2 <2>
+spec:
+  compositeTypeRef:
+    apiVersion: vshn.appcat.vshn.io/v1
+    kind: XVSHNKeycloak
+  mode: Pipeline
+  pipeline:
+    - functionRef:
+        name: function-appcat-v4-120-2 <1>
+      input:
+----
+<1> Reference to the function
+<2> Label that can be used to reference a Composition Revision in a claim or composite via a selector
+
+Once `function-appcat:v4.120.3-func` gets released and the composition changes, then there will still be the old revision that explicitly references the old function. Each claim and composite that point to that revision will reconcile against the old function, not against the new one.
+
+This emulates the same behavior as standard Composition Revisions. If at any point in the future Crossplane implements a proper support for function revision, our own solution should be sunset and instead be switched to Crossplanes solution.
+
+== High level flow
+The following flowchart shows how some additional logic in the maintenance logic for each service can enable a transition from one revision to the next.
+
+This is independent on whether we use our DIY function revisions
+
+[mermaid,flow,png]
+....
+flowchart TD
+    start[Claim/Composite created] --> test{Test OpenShift Cluster?}
+    test --> |Yes| testcluster[Always use latest revision]
+    testcluster --> stop
+    test --> |No| prod[Set revision policy to manual]
+    prod --> maint[Maintenance Cronjob triggers]
+    maint --> first{Is first maintenance for instance?}
+    first --> |Yes| autpolicy[Set revision policy to auto]
+    first --> |No| selector[Set revision selector to most current revision]
+    selector --> changes[Instance reconciles against latest function and receives changes]
+    autpolicy --> selector
+    changes --> stop
+....
+
+For production clusters the default revision policy will be set to manual. This will ensure that newly created instances will just have the most current revision. Test cluster will get the default policy automatic, so they get changes immediately. This will help detect issues.
+
+During the very first maintenance for any given instance, the maintenance job will switch the revision policy over to automatic. It will then also set the revision to the must current using a `selector`. The policy has to be set back to automatic in order for the selector to actually apply otherwise it will still keep using the old revision.
+
+=== Hotfixes
+It might be necessary that a hotfix has to be applied immediately to all instances.
+
+To handle this an additional job will be specified in the component. This job will have a parameter which contains the explicit version that should be applied to all claims or components.
```

Comment:
Can you add more about how this job is to be added? maybe in the CD GitHub pipeline where during creating a tag the pipeline understands it's a hotfix and generates the necessary job?

Reply:
This should not be automatically determined.

I'll add more how it should work.